spec.html

```diff
@@ -12148,6 +12148,21 @@ <h1>Forward Progress</h1>
     </emu-note>
   </emu-clause>
 
+  <emu-clause id="sec-async-waiters-processing-model">
+    <h1>Processing Model of Atomics.waitAsync</h1>
+    <p>At any time, an ECMAScript implementation may perform the following steps atomically:</p>
+    <emu-alg>
+      1. Let _now_ be the time value (UTC) identifying the current time.
+      1. For each WaiterList _WL_, do
+        1. Perform EnterCriticalSection(_WL_).
```

Comment:
Why isn't there a surrounding agent? This runs in some agent.

Good call on the "atomically" thing, which seems to be wrong since the critical section already ensures mutual exclusion.

Reply:
> Why isn't there a surrounding agent? This runs in some agent.

"While an agent's executing thread executes jobs, the agent is the _surrounding agent_ for the code in those jobs." Are these steps being performed as part of a job?

It seems wrong that one agent in one agent cluster would be tweaking WaiterLists in all agent clusters. I would expect that an agent can't affect any agent cluster other than its own.

Reply:
Oh agent _cluster_. Sure we can qualify it as "intra agent cluster" but that seems definitionally true. WaiterLists are associated with SAB blocks, so how could an agent have gotten WaiterLists from another agent cluster?

This is the same thing as whatever finalization registry is doing, which is admittedly handwavy but I don't really see a problem. I don't think there's any confusion there that the set of non-live objects _S_ is across agents, just like here it's not across agent clusters.

Reply:
> "While an agent's executing thread executes jobs, the agent is the surrounding agent for the code in those jobs." Are these steps being performed as part of a job?

...yes, sure?

Reply:
So is the idea something like this?
- Within each agent cluster:
  - an agent is chosen (arbitrarily?), and
  - a job is enqueued for some future time, to:
    - execute these steps with that agent as the surrounding agent;
    - and enqueue another such job.

(And the job-execution constraints guarantee that the agent won't be in any critical section when the job runs?)

Reply:
It's simpler than that. The idea is:

- Within each agent cluster:
  - choose an agent
    - run some steps in that agent, which can enqueue a Job

These steps are intended to avoid specifying a parallel watchdog thread in spec-ese, so this just says that at any time, an agent can interleave some steps (or just run them, if no other steps are running), which can then enqueue a Job for the host to run to resolve the waitAsync promise. The Job execution guarantees are for the enqueued Job that resolves the promise, not for the code that calls the host to enqueue the Job. The WL critical section takes care of synchronization for calling the host to enqueue the Job.

Reply:
You lost me there. This algorithm enqueues a Job to resolve a promise?

I was asking about the circumstances under which this algorithm (i.e., the one in "Processing Model of Atomics.waitAsync") would be executed.

Reply:
> This algorithm enqueues a Job to resolve a promise?

Yes, that's how the host integration works. The notifier agent may differ from the agent that called `Atomics.waitAsync`, and ecma262 cannot reach across agents to directly fulfill another agent's promise. So the host integration works by the notifier agent calling a host hook, which enqueues a Job in the agent that can fulfill the promise.

In the case of timeouts, there is no "notifier agent", so this "Processing Model of Atomics.waitAsync" section is intended to be maximally simple and basically says any agent at any time can check if any async waiter in its agent cluster has had its timeout expired, and if so, call the host hook.

Does that make sense?

Reply:
> any agent at any time can check if any async waiter in its agent cluster has had its timeout expired, and if so, call the host hook.

Ah, via `NotifyWaiter`, got it.

> The Job execution guarantees are for the enqueued Job that resolves the promise, not for the code that calls the host to enqueue the Job.

(That's assuming that the latter isn't itself a job, which I gather is how you're thinking about it now. But earlier, you indicated it *was* a job, so that's what I based my question on.)

So, updated question: if the algorithm of "Processing Model of Atomics.waitAsync" *isn't* part of a job, then the job-execution guarantees don't apply, so how does it satisfy `EnterCriticalSection`'s assertion that the surrounding agent is not in any critical section?

Reply:
> So, updated question: if the algorithm of "Processing Model of Atomics.waitAsync" isn't part of a job, then the job-execution guarantees don't apply, so how does it satisfy EnterCriticalSection's assertion that the surrounding agent is not in any critical section?

Yeah, that's true, there's no re-entrancy guard currently.

We can make those steps have job-like guarantees without it making a job, I guess? Another alternative formulation is perhaps to do away with the current host hook entirely and directly encode the the timeout and promise fulfillment into the "processing model". Something like,

```
At any time in an agent _agent_, when the execution context stack is empty, an implementation may
perform the following steps atomically:

1. For each WaiterList _WL_, do
  1. Enter critical section for _WL_.
  1. For each _waiter_ in _WL_ whose [[AgentSignifier]] is _agent_'s signifier, do
    1. If timeout expired, resolve the [[PromiseCapability]]. (We can do this directly since it's the right agent.)
  1. Exit the critical section for _WL_.
```

Reply:
> But then why not just say it is a job? I.e., is there something about the definition/handling of a job that needs to not apply to this algorithm?

We can. My initial hesitation was from how the HostEnqueueResolveInAgentJob hook was originally formulated. All that job does is to resolve the promise in the right agent, because an agent can only resolve its own promises. So to make the timer watchdog thing _itself_ a job, like HostEnqueueWatchdogTimerJobForSomeWaiter, we'd either need 2 jobs or something more complicated, like a requirement that the host re-enqueue the job if the timer is not yet expired. 

The HostEnqueueWatchdogTimerJobForSomeWaiter thing is also fine by me, and is probably closer to how real implementations work. Now that I've talked through it, I think that version is the cleanest on top of our existing machinery without having to modify much.

> So that 'resolve' would be something like:
Call(_waiter_.[[PromiseCapability]].[[Resolve]], *undefined*, « *"timed-out"* »).
?

Yep.

Reply:
I refactored the timeout thing to a Job and removed the "processing model" section entirely. PTAL.